library/x509write_crt.c

```diff
@@ -313,9 +313,15 @@ int mbedtls_x509write_crt_der( mbedtls_x509write_cert *ctx, unsigned char *buf,
     c = tmp_buf + sizeof( tmp_buf );
 
     /* Signature algorithm needed in TBS, and later for actual signature */
-    pk_alg = mbedtls_pk_get_type( ctx->issuer_key );
-    if( pk_alg == MBEDTLS_PK_ECKEY )
+
+    /* There's no direct way of extracting a signature algorithm
+     * (represented as an element of mbedtls_pk_type_t) from a PK instance. */
+    if( mbedtls_pk_can_do( ctx->issuer_key, MBEDTLS_PK_RSA ) )
+        pk_alg = MBEDTLS_PK_RSA;
+    else if( mbedtls_pk_can_do( ctx->issuer_key, MBEDTLS_PK_ECDSA ) )
         pk_alg = MBEDTLS_PK_ECDSA;
+    else
+        pk_alg = MBEDTLS_PK_NONE;
```

Comment:
Shouldn't we just return in the `else` case?

Reply:
Yes, that would amount to the same. I find this way a bit easier to read as we only return in one place, but I could be convinced to change it.

Reply:
Ok, so it would amount to the same because we expect `mbedtls_oid_get_oid_by_sig_alg()` to fail. Right?

Reply:
Yes

Reply:
I see. The reason why I thought that it would be good to just return here is because I am not sure whether `mbedtls_oid_get_oid_by_sig_alg()` would return in the case that `pk_alg == MBEDTLS_PK_NONE && ctx->md_alg == MBEDTLS_MD_NONE`. I am not even sure this is possible though, but `mbedtls_oid_get_oid_by_sig_alg()` does seem to have the combination as an options: https://github.com/ARMmbed/mbedtls/blob/development/library/oid.c#L405. Please correct me if I am wrong though, I did not spend long enough looking at this to figure out if `ctx->md_alg` can even be that value...

Reply:
Thanks for pointing that out. Regardless of whether this can actually occur or not, I think it's preferable to follow your initial suggestion of returning immediately, as it expresses more clearly that we are expecting to always hit either the RSA or ECDSA case - given the current code, with some time passed we would probably not remember that, wondering whether there was actually a use-case for the `MBEDTLS_PK_NONE` case.

Reply:
Done.